### ISSUE TITLE
Docs: Build code preview component

### ISSUE THREAD

**Issue by xak** [SAP/techne#723]
The site has a very basic implementation of the Code component on the website.

# Inputs
* There is a [Nunjucks macro](https://github.com/SAP/techne/blob/v2.0/docs/styleguide/macros/format.html) that wraps the code and outputs some markup. You can see it used on the [components pages](https://github.com/SAP/techne/blob/v2.0/docs/styleguide/tables.html#L31). A basic [SASS file](https://github.com/SAP/techne/blob/v2.0/docs/styleguide/styles/_code.scss) already exists.
* The visual design can be see [on Invision](https://sapse.invisionapp.com/share/DTBXJCSJW#/238944317_Docs-ComponentPage) (near the bottom). And that [PSD is on Box](https://app.box.com/folder/27696923061).

![screen shot 2017-06-13 at 5 11 31 pm](https://user-images.githubusercontent.com/424119/27106988-7705545e-505b-11e7-8626-bbd16c9fe019.png)

## Notes
* The "Note" above the code should be optional
* A JS may be available to do some magic on the markup.
* This is a standalone component (on the website) so it will take the width of it's parent.

# Outputs
Updated Code component that matches the visual design.

**Comment by saadmhybris** [comment]
After looking up some code syntax highlighter js libs, [code-prettify](https://www.npmjs.com/package/code-prettify) seems like a good option. It is being used by code.google.com and stackoverflow.com we can always use the current js lib which is [highlight js ](https://highlightjs.org) it's well documented, active and easy to impliment.

@xak any preference?

**Comment by saadmhybris** [comment]
@LeoT7508  - the red color used in the code snippet in the attached screen is not the final design right? I mean there is no code highlighting. 

are we coming up with a new color scheme or sticking with the old that looks like this below? 
<img width="469" alt="code-highlight" src="https://user-images.githubusercontent.com/15215400/27802772-78561a4c-5feb-11e7-9c13-e8b7eb7a6f05.png">

**Comment by LeoT7508** [comment]
@saadmhybris 

Yes the color you were using before will do fine for code snippets.

**Comment by joseegm** [comment]
### First thing
@xak 
**The invision link on https://github.com/SAP/techne/issues/723#issue-235708858 points to the wrong screen, again.**

**This happens a lot. Please post the PNG on the issue and don't rely only on invision links, that doesn't help, make finding information difficult and delay all the process.** 

Thank you!

**Comment by joseegm** [comment]
### Review results

1. Version number should be available for the whole component, not for individual snippets of code as it is right now.
    ![screen shot 2017-07-21 at 10 59 50](https://user-images.githubusercontent.com/22662903/28456981-935b50b0-6e04-11e7-8ff0-c364c89c083b.png)

2. No code highlighting
    ![screen shot 2017-07-21 at 11 11 21](https://user-images.githubusercontent.com/22662903/28457184-726ca72c-6e05-11e7-8a1f-25a9878b0887.png)

3. Code snippet seems wrong or truncated. It's not obvious the code can be expanded.
    ![screen shot 2017-07-21 at 11 13 06](https://user-images.githubusercontent.com/22662903/28457299-dacec642-6e05-11e7-8775-d3e035ba5d9b.png)

   **the code needs to be expanded by default**

   It also will be no problem when the tabs get implemented. More on the tabs at: https://github.com/SAP/techne/issues/694#issuecomment-304610787

4. Not sure if the "stripes" have any value here, they introduce a lot of noise. 
    The main use of this component is looking at the code.

    ![screen shot 2017-07-21 at 11 18 18](https://user-images.githubusercontent.com/22662903/28457429-639f3380-6e06-11e7-92c7-a4445c628101.png)


   **There is other reason we have the stripes?** other than aesthetics

5. Bug on the "expander", it loose the "## lines" caption once another code snippet is opened on the same page.
![screen shot 2017-07-21 at 11 23 11](https://user-images.githubusercontent.com/22662903/28457671-54252f08-6e07-11e7-9665-98b276ae04b1.png)

    **This should have been catched on the Code Review.**
    **Where is the original ticket and the PR?**

    **Note**: this may not be relevant if we address point 4.

6. Automatic indentation?
    ![screen shot 2017-07-21 at 11 30 36](https://user-images.githubusercontent.com/22662903/28458024-8066af14-6e08-11e7-9840-62915ed9e7a2.png)

    **Are the snippets indented by hand?**
    We had manual indentation on the last version and that was a mess.
    **That can be automated.**

**Comment by joseegm** [comment]
This was reviewed and put back to "in progress".

**Comment by saadmhybris** [comment]
1. The version number should be available for the whole component, not for individual snippets of code as it is right now.

We talked about this and will change it for the whole component. however, there would be cases where modifiers are introduced after different version iteration. In which case, this placement would be valid. We can discuss this when we get to that point or if it really is an issue. 

2. No code highlighting

There is a 404 error on the GitHub site which is breaking the whole jquery code. I will investigate why. This is not the case when you **pull down the branch**, therefore it was not apparent in the code review. 

3. Code snippet seems wrong or truncated. It's not obvious the code can be expanded.

This was built per design. Please check with @LeoT7508 on this. 

4. Not sure if the "stripes" have any value here, they introduce a lot of noise.
The main use of this component is looking at the code.

This was built per design. Please check with @LeoT7508 on this. 

5. Bug on the "expander", it loose the "## lines" caption once another code snippet is opened on the same page.

Probably related to the same JS error in point 2. 

6. Automatic indentation?

Indentation is not automatic, it's wrapped in a <pre> tag so the code needs to be indented properly. We had trouble with indenting by hand in the past because the code snippets were mixed with KSS code. I will investigate automatic indentation. Please let me know if you are aware of anything of the hand.

**Comment by LeoT7508** [comment]
@joseegm 

A few thoughts about a couple things here. 

#3 
This is how we initially designed this page, the code is truncated automatically to avoid having a page that scrolls forever. 

The placement of
"... 25 more lines of code    |    Show All"
was meant to work somewhat like pagination. These actions usually live on the right below and or above the area of information. This also lines up on the right with the "Copy Source Code" button, this was meant to keep all the options for the code in one area, easy to find and in one place.

#4 
Alternating row colors are meant to make long lines of text easier to read and more importantly scan for  important lines of information. 

@xak @saadmhybris  That being said, we are working with developers on this project, their feedback could helps us make a decision here about keeping or getting rid of the stripes. If they think it helps them read, scan, and find lines of code then we can keep it, if they think its a distraction then I'm totally fine dropping it. 


Overall there's a lot of inconsistencies with this page, I think the structure definitely changed and that is affecting the spacing, type, and overall flow here. The layout in the build is much different than the design we created months back. If we are going to change the structure then we have to go back and lay the design out again to make sure everything is working together.

**Comment by joseegm** [comment]
@saadmhybris 

1.  If tracking versions numbers for individual modifiers is not a hasle I'm all for it. But it should be less prominent, right now is visually too heavy.

    **Side note**: we wanted to have a way to access the old versions of the site, so maybe we don't need to directly specify by component or modifier if we make all versions available. 

    We keep the version numbers for the main component for now.

2. Okay. jQuery is very fragile. Let me know when that's working.

3. Will be addressed bellow.
4. ^^

5. Let me know when that's fixed.

6. Sure: http://jsbeautifier.org/; MIT License; https://github.com/beautify-web/js-beautify

**Comment by joseegm** [comment]
@LeoT7508 

3. I think we can prescind of hidding part of the code, when using the proposed solution back in May https://github.com/SAP/techne/issues/694

    Here for context and side by side by side (click to zoom)

    | Wireframes | Visuals |
    |---|---|
    | ![image](https://user-images.githubusercontent.com/22662903/28473213-ee594b22-6e43-11e7-994f-43d34b4a3a94.png) | ![28473278-31fe2578-6e44-11e7-93a7-e96cb5b45017](https://user-images.githubusercontent.com/22662903/28473370-87616ef8-6e44-11e7-9de6-e4065160fc8f.png) | 

    The code will be the secondary option permitting to only see the rendered components and after choosing the right one, then see the code; no need for hiding part of it.

   **The visual design for the individuals components: headers and documentation text, rendered HTML area, Code, etc are already on the visuals**. So I don't think we have to retouch those as well.

    Although the right use for the **Headers** should be there, right now they are used to display the modifier CSS classes.

4. Yes, indeed, maybe we need to have the color a little lighter, so they are there but barely visible.

To the final Leo's comment: Yes, definitely this looks different from what the wireframes or the visuals are.

If we match what we have on the wireframes with the style of the individual components from the visuals then we will have something more usable.

@saadmhybris 
**You think we can make the tabs happen?** 
It will solve a lot of the current problems

**Comment by joseegm** [comment]
### Next steps

- [ ] Implement tabs
- [ ] Put the documentation in the right place: **Bellow the Use Case Title**
- [ ] Remove the expand code -is hidden behind a tab anyway.

@saadmhybris 
This issue is only for the **Code Component**.
**Should we put this in a separate issue or keep it here for context?**

**Comment by xak** [comment]
I would like to keep the code block more prominent since these pages are important for developers. I like Leo's expandable blocks — they don't take up much room but are easily visible. 

Viewing the code at the same time as the styled example can be useful for comparison ... and hiding the code behind a tab requires an extra click — a developer just wants the code.

The tabs don't add much UX value and they add some repetitive visual noise seeing them over and over again on the page.

The solution implemented satisfies the [requirements](https://github.com/SAP/techne/issues/723#issue-235708858) for this issue.

**Comment by joseegm** [comment]
Argument about the second click doesn't hold,  the code is visible with only 3 lines, **second click still needed to expand the code.**

Visual repetition, that can solved: **We can replace it with icons.**

Instead of tabs we can expand the code on click, that we it is shown side by side. That will be a good solution.

Just for context:
- Material: icon and expandable
     ![img_0010](https://user-images.githubusercontent.com/22662903/28477217-473c8f92-6e53-11e7-8564-b4c717cfcbe1.PNG)

    ![img_0011](https://user-images.githubusercontent.com/22662903/28477230-55ac083c-6e53-11e7-8abc-997eedd9e19c.PNG)

- Fabric: icon and label expandable
    ![img_0012](https://user-images.githubusercontent.com/22662903/28477264-78632b44-6e53-11e7-8d20-15001539771e.PNG)

**Remember, the developer first look visually which component type to use, once found, then look at the code, not the other way around**. The developer not just want the code first.

The UX value of the tabs or expandable hiding the code is: support the above mentioned behavior. 
Having the code visible all the time doesn't support the visual search.

**We can go with an icon and expanding the code, that way the user see both after deciding which type he needs**

**Comment by LeoT7508** [comment]
@joseegm  @xak 

I think the big question here is why hide the code if we don't have to? If we can show both why wouldn't we?

**Comment by joseegm** [comment]
The Use case for this page is as already mentioned https://github.com/SAP/techne/issues/723#issuecomment-317080183

Going to describe the flow in more detail. Here are the steps and the goals of the user:
1. Developer get the component from a wireframe: Let's say it is the card with image.
2. Go to the Cards page and **search visually** for the component.
    - Our goal here is to **assist the visual search**.
3. Find the indicated component type and **only then want to see the code**
    - Secondary goal is to get the code.

So to assist this flow:
1. Make the visual search faster: **by hiding the code**, which is only important after the visual search is concluded.
2. Access the code with one click.

With that in mind: **The expandable code** will support the use case.

@LeoT7508 
That we have a piece of information available doesn't mean we should show it all the time. Only when it support the use case.

Think now is all clear what the use case and the features of this page are.

All that is reflected on the wireframes: https://github.com/SAP/techne/issues/723#issuecomment-317054252

**Comment by xak** [comment]
@joseegm This needs to be a separate issue, of course, but I believe there are other use cases here to consider:

**A developer wants to understand what modifications and states are available for a component and how that affects the rendering.**

**A developer sees a class name in a project source code and wants to learn more about that component. The developer comes to the component page to search for that class name.**

We should expect teams to be using this library for years, bringing in new developers and on boarding them to the project. Some teams will create a single UI and never touch it again ... but that is not the only way.

Good UX takes into account the various ways people use the page and attempts to find the best solution. Some use cases may seem like a priority but often user testing shows that not to be true. I don't know here what the best solution is here but I hope we can broaden our thinking.

As with anything else, we can prioritize quick hits to get something out there, we just have to align with design and development to make sure there is no confusion about competing visions.

**Comment by LeoT7508** [comment]
@joseegm @xak 

From a simplicity and flow perspective I would vote to keep the code exposed. It solves both use cases at once and sticks closer to the original idea and design. It reduces any confusion, makes it easier to find, does away with a click, and keeps the whole idea simpler. 

Either way we should create a new design document to reference.

**Comment by joseegm** [comment]
This will be put in a separate issue to assess and select all uses cases and from there revisit the Component page structure, etc.

**Comment by joseegm** [comment]
Issue created #842 

The Code Component is already merged. It will be revisited once #842 is finished.

Putting this to done and closing it.
